### ISSUE TITLE
Suggested alternative usage, more similar to Google Mock

### ISSUE THREAD

**Issue by ugoren** [hjagodzinski/C-Mock#9]
Hello,
I was shown C-Mock and find it a very useful extension to Google Mock.
When using it, I was a little annoyed by the need to define a separate mock class for each mocked function.
I believe, based on reading the code and my experience in messing with Google Mock under the hood, that better usage is possible, and I can implement it.

Here's my suggestion:

In the header file, declare a mock class, similarly to Google Mock, where each mocked function is a method:
```
class FooFunctionMock : FunctionMocker<FooFunctionMock>
{
public:
     MOCK_FUNCTION2(foo, int, (int, int));
     MOCK_FUNCTION1(bar, int, (int));
};
```
The `FunctionMocker` class will allow access to the `FooFunctionMock` instance (similarly to the `instance` pointer).
`MOCK_FUNCTIONn` will be similar to `MOCK_METHODn`, but will declare a `real_fname` member and `call_fname` method.

In the mock source file, use `IMPLEMENT_FUNCTION_MOCK` similarly to today:
```
IMPLEMENT_FUNCTION_MOCK2(FooFunctionMock, foo, int(int, int)); 
IMPLEMENT_FUNCTION_MOCK1(FooFunctionMock, bar, int(int));
```
Unlike the current macro, it won't implement a constructor and destructor, just `real_fname` and `call_fname`. So there's no problem in using it multiple times with the same class name.
I guess we'll need to rename it, to avoid conflict with the existing macro.

Expectations will be set using `EXPECT_CALL`, just like any Google Mock class.

**Comment by hjagodzinski** [comment]
Hi @ugoren,

Sorry for late response. I'm glad you find C-Mock a useful too. I like your idea. Apart from avoiding creating class for each mocked function I see one more advantage. You can logically organize mocked functions by separate mock classes - let it be a set of functions coming from a single library put into single mock class.

**Comment by ugoren** [comment]
Thanks.
I already started implementing it, I'll update my clone tomorrow.
It works almost as planned, with one difference - I need to define a static
member per class, and this requires another IMPLEMENT_ macro called one per
class.

On Feb 3, 2018 11:03 AM, "Hubert Jagodziński" <notifications@github.com>
wrote:

Hi @ugoren <https://github.com/ugoren>,

Sorry for late response. I'm glad you find C-Mock a useful too. I like your
idea. Apart from avoiding creating class for each mocked function I see one
more advantage. You can logically organize mocked functions by separate
mock classes - let it be a set of functions coming from a single library
put into single mock class.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/hjagodzinski/C-Mock/issues/9#issuecomment-362792084>,
or mute the thread

**Comment by ugoren** [comment]
My clone - https://github.com/ugoren/C-Mock - has a working implementation.
Usage is actually simpler than I suggested - no need for the `MOCK_FUNCTION` macros. You can see the added tests for the exact usage.

The new IMPL macro is named `IMPLEMENT_FUNCTION_MOCK_N$i` - I used `_N` to distinguish from the normal macro, but it's obviously not a good name. I'll try to come up with a better name.

**Comment by hjagodzinski** [comment]
Sorry I'm pretty busy lately.

I see you finally decided on a `CMOCK_REDIRECT_TO_MOCK$i`:) I like you used CMOCK_ as a prefix. Why did not think about it? ;)

I have one suggestion though. If I don't miss anything you can define `FunctionMocker` without a template declaration. This way you will hide an implementation detail making a function mocker creation a bit simpler. In each place you use `cmock_get_instance` you will simply cast it to the actual type.

**Comment by ugoren** [comment]
Hello,

I don't see how to avoid the templatization of FunctionMocker. It allows
the static pointer to be separate for each inheriting class, rather than a
single instance pointer across all mock classes.

Regarding names - I'm really not sure I have them right. They can probably
be better.
1. I switched from IMPLEMENT_FUNCTION_MOCK to CMOCK_REDIRECT_TO_MOCK when I
realized that it's no longer paired to DECLARE_FUNCTION_MOCK, It's actually
kind-of paired to Google's MOCK_METHOD. I think the name is a bit long, and
contains the word "mock" twice. Perhaps CMOCK_IMPLEMENT_METHOD would be
better?
2. FunctionMocker - I don't like this one very much. It doesn't have a
cmock prefix or namespace, and its name is not so much related to what it
does. It's more of a SingletonMock.
3. cmock_get_instance - I think this one is OK. It's not used externally,
so the only thing that matters is that it will be unlikely to collide with
the user's methods.

**Comment by hjagodzinski** [comment]
Hi,

You're absolutely right on templarization. For some reason I was still subconsciously convinced the class is each time generated by a macro.

1. I don't see it as a problem that the macro contains the word mock two times (the first occurrence is simply the project name). I have an idea to name the macro like `CMOCK_MOCK_FUNCTION`. This name in a way reflects the Google Mock's naming convention.

2. To me `FunctionMocker` says what the class is used for whereas `SingletonMock` says how it's implemented internally.

3. I agree.

Today I'm going for holidays and won't be available for around four weeks. We can come back to this when I'm back if it's fine with you.

**Comment by ugoren** [comment]
Hi,
Whatever names you want are fine with me.
I already started using these modifications myself, and intend to spread it
in my company. I prefer to use the exact same names you'll eventually use,
to save me the bother of changing later. But I don't worry too much,
because I don't think it will be used very extensively so quickly, so if we
need to change it won't be so hard.
Meanwhile, I'll go with CMOCK_MOCK_FUNCTION and FunctionMocker.

Enjoy your holidays.
I'll have my fork updated by the time you return.

**Comment by ugoren** [comment]
Hello Hubert,

I hope you had a good vacation.

I just made one small change - I realized that FunctionMocker was not a
good choice for a name, as it conflicts with Google Mock's ::testing:
FunctionMocker . It worked well, until I did "using namespace testing" in
one test.
So I renamed it to CmockFuncMocker.

If that's OK, then I see nothing blocking us from integrating all this into
your branch.

Thanks,
Uri.
